### ISSUE TITLE
Service Workers often fail to fetch data on web platforms

### ISSUE THREAD

**Issue by skymen** [Scirra/Construct-bugs#7013]
## Problem description

Very often, when publishing a Construct game on web portals, an issue happens where Construct just fails to load the game from the service worker.

This does not happen all the time and it's very inconsistent but it's happening often enough for it to be a problem.

In addition, service workers serve zero purpose on these platforms since the game can't start loading until the platform itself has loaded, making the game "online" regardless.

I have done some tests specifically with Poki engineers and we found that Service Workers are not only useless but actively disrupt the request cache of Chrome and not only makes the game fail to load sometimes, but it also makes the game load SLOWER on average.

## Attach a .c3p

N/A

## Steps to reproduce

N/A

## Observed result

Here is another bug report talking about the same issue but happening on Itch.io
https://github.com/Scirra/Construct-bugs/issues/6223

And here is an error from another Construct game on Poki this time:
![image](https://user-images.githubusercontent.com/11213825/236833119-ad3feb0f-cb60-4886-a67a-dfe30bfb8980.png)

This is not a platform specific issue, as far as I'm aware, it happens on almost all of the major web game portals out there.

## Expected result

These issues should not happen

## More details
The solution IMO is to just provide a way to disable service workers entirely. It should be very easy to do since the exported game has a way for you to manually remove the service worker. This is very tedious to do however since it requires doing an additional pass on the game files after every export.

A simple toggle under "Advanced" tab would work wonders, and all it would do is remove the registerSW.js file from the exported zip.

Doing this would also fix a lot of issues related to the "Updating the game" part since Chrome request cache also already does this more cleanly. A lot of web platforms serve different game versions on different URLs to prevent the Service Worker from doing this as well, and I'm not even speaking about all the issues this causes when making multiplayer web games.

Service workers have some benefits, but genuinely, on web platforms it does no good, and causes many issues that can't be worked around other than removing it.

**Affected browsers/platforms:** Chrome, all platforms. Haven't thoroughly tested on Firefox and Safari

**First affected release:** Construct 2, this has always been an issue

## System details

<details><summary>View details</summary>

N/A

</details>

**Comment by AshleyScirra** [comment]
I'm afraid as per the bug report guidelines there's no way we can investigate without a sample project and steps to reproduce. Further this seems to conclude with a feature request, so that should be posted to the suggestions platform instead.

There's very little to go on but from the information I have available I would say:

1) issue #6223 involved the server returning a 403 error response, which is an issue with the server, not Construct
2) the only issue I can see in the console log in this issue is a 404 error response, which is an issue with the server, not Construct
3) the Service Worker is coded to be fail-safe so even if it fails to cache files for offline use, the project should continue to load normally
4) sometimes network connection errors happen and they can prevent the game from loading, but that is not itself a problem with Construct, and adding an option to turn off the Service Worker won't help you there, since the problem happened in the network and not the Service Worker. The fact the console mentions service workers does not actually prove it is the cause of the problem, it's probably just where the error got logged. So I think that could well be a red herring.

I don't think there's any evidence from this that the Service Worker is actually the problem, but there's no way to further investigate, so closing. I'd advise to at least look in the Network tab of dev tools on a failed load to try to find any failed requests; if the network request failed, that is more likely a problem with the Internet connection or the server than with Construct.

**Comment by skymen** [comment]
Listen, what you say makes a lot of sense, however from my own experience, removing service workers have fixed the issues completely very consistently.

I'm not talking about a one off thing, or a thing that happened by chance. Over the past 5 years, I have had maybe 8 or 9 project that had similar issues and all of them without fail were fixed by removing service worker.

I trust that you know how service workers work a lot more than me and maybe what you say makes a lot of sense, but that does not match my experience and the experience of fellow web developers.

I think the 404/403 responses are the red herring which lead you to believe this is a server issue, when I think it's actually a misreported Chrome issue that can be worked around by disabling service workers.

> the Service Worker is coded to be fail-safe so even if it fails to cache files for offline use, the project should continue to load normally

I think you misunderstand what I'm reporting about: I am not reporting about the service worker failing to cache files for offline use. I'm talking about the service worker successfully caching them, but failing to load them from the cache, and halting the project's load because of that.

It seems though that when this happens, the next time the project loads, it's loaded from the server and it works fine, but then the cycle repeats as soon as the service worker caches them again.

**Comment by erikdubbelboer** [comment]
The 404 you are seeing is the service worker requesting a path that doesn't exist. There is no content on `/`, only on `/index.html`. Why would the service worker try to request this path? Does it maybe expect `/` to be routed to `/index.html` automatically? Could it cause any issues when this isn't done?

**Comment by AshleyScirra** [comment]
Some web servers serve the same content for `/` and `/index.html`. If it offline caches just one, then it will fail to load offline on the other URL, which has been problematic in the past. So if it loads on `index.html`, it also tries to cache `/` in an effort to make sure it still works offline if the user later visits the `/` URL while offline. This also matches the default configuration for most web servers. I guess it perhaps should treat that as an optional fetch and not fail if it returns 404, but this is unrelated to the issue mentioned here - as I said, if any requested resource returns 404, the SW is fail safe and the project continues working fine.

**Comment by skymen** [comment]
I was about to create a feature request as per your request only to find I had already done it a few months ago and forgot about it:
https://construct23.ideas.aha.io/ideas/C23-I-198

I think the issue @erikdubbelboer talked about is interesting and might be worth a look. 

About the subject at hand, if you would rather treat this as a feature request and not as a bug fix, feel free to do so and in that case you can consider the specific issue I reported closed
